smart_open/s3.py

```diff
@@ -614,7 +614,9 @@ def _accept_all(key):
 
 
 def iter_bucket(bucket_name, prefix='', accept_key=None,
-                key_limit=None, workers=16, retries=3):
+                key_limit=None, workers=16, retries=3,
```

Comment:
I think passing in these parameters as three separate things is a bit ugly. This is especially true because we now do this in like 3-4 different functions.

Is it possible to pass around a boto3.Session instead?

Reply:
If we pass a Session, we can also close #366, which will be a big win.

Reply:
Per https://github.com/boto/botocore/issues/1246 it would appear that it is not recommended that we can pass boto3.Session around. (Whatever recommended means...)

I totally agree passing separate things around is ugly, but don't know if we should take this risk.

Reply:
> It is recommended to create a resource instance for each thread / process in a multithreaded or multiprocess application rather than sharing a single instance among the threads / processes.

OK, I see. We need to pass session information across a thread/process boundary. In that case, we shouldn't pass around a single session instance.

Do this instead:

```python
def iter_bucket(..., **session_kwargs):
    """
    session_kwars: dict, optional
        Keyword arguments to pass when creating a new session.
        For a list of available names and values, see:
        https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html#boto3.session.Session
    """
    ...

def list_bucket(..., **session_kwargs):
    client = boto3.client('s3', **session_kwargs)

```